File: day02/haskell/src/Main.hs
module Main (main) where

main :: IO ()
main = do
  input <- readFile "../input"
  putStrLn $ "First answer: " ++ (show $ firstAnwer input)
  putStrLn $ "Second answer: " ++ (show $ secondAnswer input)
  where
    firstAnwer = sum . map score . lines
    secondAnswer = sum . map (score . translate) . lines
    
    score "A X" = 1 + 3
    score "A Y" = 2 + 6
    score "A Z" = 3 + 0
    score "B X" = 1 + 0
    score "B Y" = 2 + 3
    score "B Z" = 3 + 6
    score "C X" = 1 + 6
    score "C Y" = 2 + 0
    score "C Z" = 3 + 3
    score _ = undefined
    
    translate "A X" = "A Z"
    translate "A Y" = "A X"
    translate "A Z" = "A Y"
    translate "B X" = "B X"
    translate "B Y" = "B Y"
    translate "B Z" = "B Z"
    translate "C X" = "C Y"
    translate "C Y" = "C Z"
    translate "C Z" = "C X"
    translate _ = undefined

File: day02/rust/src/main.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("../input").expect("input file missing");

    let first_answer = input.lines()
        .map(score)
        .sum::<u32>();
    println!("First answer: {first_answer}");

    let second_answer = input.lines()
        .map(|game| score(translate(game)))
        .sum::<u32>();
    println!("Second answer: {second_answer}");
}

fn score(game: &str) -> u32 {
    match game {
        "A X" => 1 + 3,
        "A Y" => 2 + 6,
        "A Z" => 3 + 0,
        "B X" => 1 + 0,
        "B Y" => 2 + 3,
        "B Z" => 3 + 6,
        "C X" => 1 + 6,
        "C Y" => 2 + 0,
        "C Z" => 3 + 3,
        _ => panic!("unknown game: {game}"),
    }
}

fn translate(game: &str) -> &str {
    match game {
        "A X" => "A Z",
        "A Y" => "A X",
        "A Z" => "A Y",
        "B X" => "B X",
        "B Y" => "B Y",
        "B Z" => "B Z",
        "C X" => "C Y",
        "C Y" => "C Z",
        "C Z" => "C X",
        _ => panic!("unknown game: {game}"),
    }
}
